src/utils/url.js

```diff
@@ -92,64 +94,32 @@ export const tryParseUrl = (url: string, base?: string | URL): URL | void => {
 };
 
 /**
- * Test if the given URL string resolves on the realm, with realm as base.
- *
- * DEPRECATED because URL strings are complicated.  Callers should construct
- *   a URL object before this point, and use the same URL object both for
- *   asking any questions like this one and for subsequently making any
- *   network requests.
- *
- * This returns true just if `new URL(url, realm)` gives a URL that is
- * within the same Zulip realm.
- *
- * This performs a call to `new URL` and therefore may take a fraction of a
- * millisecond.  Avoid using in a context where it might be called more than
- * 10 or 100 times per user action.
+ * Test if the given URL is within the given realm.
  */
-// TODO: Take a URL object instead of a string, and remove warnings in jsdoc.
-export const isUrlOnRealm = (url: string, realm: URL): boolean => {
-  const parsedUrl = tryParseUrl(url, realm);
-  return parsedUrl ? parsedUrl.origin === realm.origin : false;
-};
+export const isUrlOnRealm = (url: URL, realm: URL): boolean => url.origin === realm.origin;
 
-const getResourceWithAuth = (uri: string, auth: Auth) => ({
-  uri: new URL(uri, auth.realm).toString(),
+const getResourceWithAuth = (url: URL, auth: Auth) => ({
+  uri: url.toString(),
   headers: getAuthHeaders(auth),
 });
 
-const getResourceNoAuth = (uri: string) => ({
-  uri,
+const getResourceNoAuth = (url: URL) => ({
+  uri: url.toString(),
 });
 
-export const getResource = (
-  uri: string,
-  auth: Auth,
-): {| uri: string, headers?: {| [string]: string |} |} =>
-  isUrlOnRealm(uri, auth.realm) ? getResourceWithAuth(uri, auth) : getResourceNoAuth(uri);
-
-export type Protocol = 'https://' | 'http://';
-
-const protocolRegex = /^\s*((?:http|https):\/\/)(.*)$/;
-
-// Split a (possible) URL into protocol and non-protocol parts.
-// The former will be null if no recognized protocol is a component
-// of the string.
-//
-// Ignores initial spaces.
-/** PRIVATE -- exported only for testing */
-export const parseProtocol = (value: string): [Protocol | null, string] => {
-  const match = protocolRegex.exec(value);
-
-  if (match) {
-    const [, protocol, rest] = match;
-    if (protocol === 'http://' || protocol === 'https://') {
-      return [protocol, rest];
-    } else {
-      throw new Error(`impossible match: protocol = '${escape(protocol)}'`);
-    }
-  }
-  return [null, value];
-};
+/**
+ * From a URL string and an Auth object, give React Native's ImageSource type.
+ *
+ * This is what React Native's `Image` component expects for its `source`
+ * prop: https://reactnative.dev/docs/0.67/image#imagesource
+ *
+ * It's also what react-native-photo-view is documented as expecting for its
+ * `source` prop: https://github.com/alwx/react-native-photo-view#properties
+ *
+ * > source | Object | same as source for other React images
+ */
+export const getResource = (url: URL, auth: Auth): ImageSource =>
```

Comment:
Similarly I think cleanest to have this start taking URL objects as its own NFC commit -- just this function, its tests, and its call site which starts calling `tryParseUrl` for itself.

That way, the lightbox-related commit is focused on the lightbox's own code (plus the spot where we navigate to it), and isn't changing shared lower-level layers like the API bindings or `src/utils/url.js` where one has to think about what else the changes might affect.